src/coreclr/vm/jitinterface.cpp

```diff
@@ -2106,20 +2106,10 @@ static GetTypeLayoutResult GetTypeLayoutHelper(
     parNode.size = pMT->GetNumInstanceFieldBytes();
     parNode.numFields = 0;
     parNode.type = CorInfoType::CORINFO_TYPE_VALUECLASS;
-    parNode.hasSignificantPadding = false;
 
     EEClass* pClass = pMT->GetClass();
-    if (pClass->IsNotTightlyPacked() && (!pClass->IsManagedSequential() || pClass->HasExplicitSize() || pClass->IsInlineArray()))
-    {
-        // Historically on the JIT side we did not consider types with GC
-        // pointers to have significant padding, even when they have explicit
-        // layout attributes. This retains the more liberal treatment and
-        // lets the JIT still optimize these cases.
-        if (!pMT->ContainsPointers() && pMT != g_TypedReferenceMT)
-        {
-            parNode.hasSignificantPadding = true;
-        }
-    }
+    parNode.hasSignificantPadding = pClass->IsNotTightlyPacked() && (pClass->HasExplicitFieldOffsetLayout() || pClass->HasExplicitSize());
```

Comment:
`IsNotTightlyPacked` condition is missing in the VM version of the check.

What does the JIT do when it sees `hasSignificantPadding` set for a structure with no padding?

Reply:
> What does the JIT do when it sees hasSignificantPadding set for a structure with no padding?

It's ignored, so it doesn't really matter. I don't think the managed type system has a similar easy to access property. I can remove it from the VM version.

Reply:
> It's ignored

Well, it's ignored except for the quirk being removed in #88130.